File: Retos/Reto #4 - PRIMO, FIBONACCI Y PAR [Media]/python/dPenedo.py
# # Reto #4: PRIMO, FIBONACCI Y PAR
# #### Dificultad: Media | Publicación: 23/01/23 | Corrección: 30/01/23

# ## Enunciado

# ```
# /*
#  * Escribe un programa que, dado un número, compruebe y muestre si es primo, fibonacci y par.
#  * Ejemplos:
#  * - Con el número 2, nos dirá: "2 es primo, fibonacci y es par"
#  * - Con el número 7, nos dirá: "7 es primo, no es fibonacci y es impar"
#  */
# ```
# #### Tienes toda la información extendida sobre los retos de programación semanales en **[retosdeprogramacion.com/semanales2023](https://retosdeprogramacion.com/semanales2023)**.

# Sigue las **[instrucciones](../../README.md)**, consulta las correcciones y aporta la tuya propia utilizando el lenguaje de programación que quieras.

# > Recuerda que cada semana se publica un nuevo ejercicio y se corrige el de la semana anterior en directo desde **[Twitch](https://twitch.tv/mouredev)**. Tienes el horario en la sección "eventos" del servidor de **[Discord](https://discord.gg/mouredev)**.
import math

def es_primo(numero):
    for n in range (2, numero):
        if numero % n == 0:
            return False
    return True

def es_fibonacci(numero):
    phi = 0.5 + 0.5 * math.sqrt(5.0)
    a = phi * numero
    return numero == 0 or abs(round(a) - a) < 1.0 / numero


while True:
    numero_introducido = int(input("Introduzca un número para realizar comprobaciones \n"))
    caracteristicas = []

    if (numero_introducido % 2 == 0):
        caracteristicas.append("Par")
    else:
        caracteristicas.append("Impar")
    if (es_fibonacci(numero_introducido)):
        caracteristicas.append("Fibonacci")
    else:
        caracteristicas.append("No Fibonacci")
    if (es_primo(numero_introducido)):
        caracteristicas.append("Primo")
    else:
        caracteristicas.append("No Primo")
    print(f"el número {numero_introducido} es: {caracteristicas[0]}, {caracteristicas[1]} y {caracteristicas[2]}")
    caracteristicas = []

File: Retos/Reto #4 - PRIMO, FIBONACCI Y PAR [Media]/javascript/dPenedo.js
// Escribe un programa que, dado un número, compruebe y muestre si es primo, fibonacci y par.
// Ejemplos:
// - Con el número 2, nos dirá: "2 es primo, fibonacci y es par"
// - Con el número 7, nos dirá: "7 es primo, no es fibonacci y es impar"

function isEven(numberToCheck) {
    return (numberToCheck % 2 === 0);
}

function isPerfectSquare(numberToCheck) {
    let square = parseInt(Math.sqrt(numberToCheck));
    return (square * square == numberToCheck);
}

function isFibonacci(numberToCheck) {
    let fib1 = 0;
    let fib2 = 1;

    while (fib1 <= numberToCheck) {
        if (fib1 === numberToCheck) {
            return true;
        }
        let temp = fib1 + fib2;
        fib1 = fib2;
        fib2 = temp;
    }
}

function isPrime(numberToCheck) {
    if (numberToCheck > 1) {
        for (let i = 2; i < numberToCheck; i++) {
            if (numberToCheck % i == 0) {
                return false;
            }
        }
        return true;
    }
    return false;
}


function numberChecker(numberToCheck) {
    let evenText = isEven(numberToCheck) ? "even" : "odd";
    let fibonacciText = isFibonacci(numberToCheck) ? "Fibonacci" : "not Fibonacci";
    let primeText = isPrime(numberToCheck) ? "prime" : "not prime";

    console.log(`Your given number: ${numberToCheck}, is ${evenText}, ${fibonacciText} and ${primeText}`);
}

numberChecker(1);
numberChecker(2);
numberChecker(5);
numberChecker(7);
numberChecker(12);